### PR TITLE
#1255 ReverseMapping should not apply automatically.

### PR DESCRIPTION
**Pull request by sjaakd**
The problem reproduces by means of the provided test case. However, it seems to be deliberate functionality, at one time introduced after issue #72.

It is possible to include a `source` name besides `target` in the `@Mapping`. Also, ignore should be applied when there is a target property with the same name in the source. The testcases in `org.mapstruct.ap.test.ignore.IgnorePropertyTest` start to fail once the fix in the second [commit](https://github.com/sjaakd/mapstruct/commit/9db22395ae3b354efa27ff5ac8bdd4065c0d50f1) is in place. 

@gunnarmorling , @agudian , @filiphr : do we still want this functionality? Seems to me it is more straightforward and better explainable to 'ignore' the `ignore=true` in the reverse mapping just as we do for expressions and constants. But that would be breaking existing functionality.

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/Mapping.java [hunk 1]
```diff
@@ -398,16 +398,16 @@ private boolean hasPropertyInReverseMethod(String name, SourceMethod method) {
...
         }
 
         // should only ignore a property when 1) there is a sourceName defined or 2) there's a name match
-        if ( isIgnored ) {
-            if ( sourceName == null && !hasPropertyInReverseMethod( targetName, method ) ) {
-                return null;
-            }
-        }
+//        if ( isIgnored ) {
+//            if ( sourceName == null && !hasPropertyInReverseMethod( targetName, method ) ) {
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/Mapping.java]: so this code seems to be specifically there to test the conditions I mentioned.

**Comment by agudian** [comment]
I’m for doing it as we discussed, not reverse the ignore... and document that change in the release notes 🙂

**Comment by sjaakd** [comment]
> I’m for doing it as we discussed, not reverse the ignore... and document that change in the release notes 🙂

I'm myself also in favor. But perhaps we should explicitly exclude the combination of `source-target-ignore` (just as I think we do for `source-target-constant` and `source-target-expression`). And we should adapt the `org.mapstruct.ap.test.ignore.IgnorePropertyTest`

**Comment by filiphr** [comment]
I have another idea. Why not reverse the ignore if and only if there is a source parameter.

So when we have:

```java
@Mapping(target = "target", source = "source", ignore = true)
```

Then we are going to produce a reverse mapping. However, if we have:

```java
@Mapping(target = "target", ignore = true)
```

Basically, the second point on line 405 will not longer hold. 

Then we won't reverse it. I think that that way one can configure what they want to reverse, i.e. if the reverse should not be inversely inherited then they should not write a source parameter, but if they want to inherit the ignore to the reverse method, than they also need to provide a source parameter.

Theoretically, your reproducer should be applied, because you have both `target` and `source`. However, in the issue there is no source, so it won't be used. and it should work as it should (I think 😄 )

**Comment by sjaakd** [comment]
Its both: name based mapping and/or presence source parameter in mapping

**Comment by filiphr** [comment]
>Its both: name based mapping and/or presence source parameter in mapping

Sorry I didn't understood you well? What is both?

Btw. I also forgot, we would need to update the Javadoc on the `MappingInheritanceStrategy#AUTO_INHERIT_FROM_CONFIG` as well. See the original message in the issue for that. The text there is not completely correct.

**Comment by sjaakd** [comment]
I meant, if @Mapping only has a target and ignore =true but source and target property name are the same, the reverse mapping is ignored as well. This is what's actually happening in the reported problem.

Verstuurd vanaf mijn iPhone

**Comment by filiphr** [comment]
Now I understand you. Yes that is what I meant as well. When the `source` and `target` name are same and the user wants to ignore in both sides then they have to write `@Mapping(source = "field1", target = "field1", ignore = true)`.

**Comment by sjaakd** [comment]
So: to If I understand you correctly: currently  @Mapping(source = "field1", ignore = true) 
Is sufficient when name are the same and you want to make this explicit like:  @Mapping(source = "field1", target = "field1", ignore = true). 
As well. @agudian and myself want to remove the option completely.

> Op 18 aug. 2017 om 23:56 heeft Filip Hrisafov <notifications@github.com> het volgende geschreven:
> 
> @Mapping(source = "field1", target = "field1", ignore = true)

**Comment by filiphr** [comment]
Yes I think that making it explicit like `@Mapping(source = "field1", target = "field1", ignore = true)` might be more clearer for users, they might expect something like this to work. Whereas with `@Mapping(target = "field1", ignore = true)`, one would expect that it is ignored only in one direction.

Basically what I am saying is to remove the `!hasPropertyInReverseMethod( targetName, method )` check.

However, if you and @agudian think that it is better to completely remove that go for it.

**Comment by sjaakd** [comment]
> Yes I think that making it explicit like @Mapping(source = "field1", target = "field1", ignore = true) might be more clearer for users, they might expect something like this to work. Whereas with @Mapping(target = "field1", ignore = true), one would expect that it is ignored only in one direction.

If you extrapolate on that thought: why not support `@Mapping(source = "field1"", ignore = true)` as well, so that a field is mapped in the forward mapping and suppressed in the reverse mapping. That would also cover fields which are not present in the source mappings like database keys. 

The only problem for this: we attached everything on the target mapping. Therefor defining 2 mappings (forward and reverse) in the shared mapping configuration (in which optionally the reverse mapping inherits everything from the forward mapping and particular fields are overriden) might in the end be more clear. Don't know if we have limitations on that though.

**Comment by filiphr** [comment]
You are right, defining a specific mapping (that would inherit all except `ignore` from the forward mapping) might be clearer.

So if we handle the reversing like you and @agudian are proposing and mention this in the release notes it should be fine.

I think that we would need some tests as well to make sure that this constellation works.

**Comment by sjaakd** [comment]
still to do: finish discussion in #1255 
Opinion @gunnarmorling ?

**Comment by sjaakd** [comment]
Need to check the (failing) integration. Otherwise its ready for review I reckon.

File: processor/src/test/java/org/mapstruct/ap/test/inheritfromconfig/InheritFromConfigTest.java [hunk 2]
```diff
@@ -141,6 +141,35 @@ public void explicitInheritedMappingIsAppliedInReverseDirectlyFromConfig() {
     }
 
     @Test
+    @IssueKey( "1255" )
+    @WithClasses({ CarMapperReverseWithAutoInheritance.class, AutoInheritedReverseConfig.class } )
+    public void autoInheritedMappingIsAppliedInReverseDirectlyFromConfig() {
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/inheritfromconfig/InheritFromConfigTest.java]: I think that we need to add the reverse method to the `CarMapperReverseWithAutoInheritance` and validate that the inheritance is not inherited

File: processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java [hunk 3]
```diff
@@ -446,41 +446,48 @@ private void mergeInheritedOptions(SourceMethod method, MapperConfiguration mapp
...
+                        true,
+                        method,
+                        messager,
+                        typeFactory );
+                }
+                else if ( applicableReversePrototypeMethods.size() > 1 ) {
+                    messager.printMessage(
+                        method.getExecutable(),
+                        Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
+                        Strings.join( applicablePrototypeMethods, ", " ) );
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java]: This should be `applicableReversePrototypeMethods` instead of `applicablePrototypeMethods`. In any case, I think it would be better if we can extract it into a method. Something like:

```java
private void applyInheritedOptions(SourceMethod method, MappingOptions mappingOptions,
        List<SourceMethod> applicableMethods, boolean isInverse) {
        if ( applicableMethods.size() == 1 ) {
            mappingOptions.applyInheritedOptions(
                first( applicableMethods ).getMappingOptions(),
                isInverse,
                method,
                messager,
                typeFactory
            );
        }
        else if ( applicableMethods.size() > 1 ) {
            messager.printMessage(
                method.getExecutable(),
                isInverse ? Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH :
                    Message.INHERITCONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
                Strings.join( applicableMethods, ", " )
            );
        }
    }
```

And we just call it with the appropriate parameters in the ifs. The logic is the same with the only difference in the methods and the `isInverse`

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 4]
```diff
@@ -47,6 +47,7 @@
     PROPERTYMAPPING_DUPLICATE_TARGETS( "Target property \"%s\" must not be mapped more than once." ),
     PROPERTYMAPPING_EMPTY_TARGET( "Target must not be empty in @Mapping." ),
     PROPERTYMAPPING_SOURCE_AND_CONSTANT_BOTH_DEFINED( "Source and constant are both defined in @Mapping, either define a source or a constant." ),
+    PROPERTYMAPPING_SOURCE_AND_IGNORE_BOTH_DEFINED( "Source and ignore are both defined in @Mapping, make explicit in reverse mapping when the intent is to ignore the reverse mapping." ),
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: This is not used. Did you forget to push some things maybe 😄 ?

File: processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousCombinationOfSourceAndIgnoreMapper.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.mapstruct.Mapping;
+import org.mapstruct.Mappings;
+import org.mapstruct.factory.Mappers;
+
+/**
+ *
+ * @author Sjaak Derksen
+ */
+@Mapper
+public interface ErroneousCombinationOfSourceAndIgnoreMapper {
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousCombinationOfSourceAndIgnoreMapper.java]: This is also not used. Linked with the message I think

**Comment by sjaakd** [comment]
> Maybe you added them to the copied sources by mistake

uhh. Yep..

**Comment by sjaakd** [comment]
I'll also add some extra test cases.

File: processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java (hunk 3 above)
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java]: I don't understand what you mean. The calling side would look like:

```java
else if ( inheritanceStrategy.isAutoInherit() ) {

            if ( inheritanceStrategy.isApplyForward() ) {
                applyInheritedOptions( method, mappingOptions, applicablePrototypeMethods, false );
            }

            if ( inheritanceStrategy.isApplyReverse() ) {
                applyInheritedOptions( method, mappingOptions, applicableReversePrototypeMethods, true );
            }
        }
```
Which is exactly the same. With the only difference those code that is the same is in a method now 😄

File: processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousCombinationOfSourceAndIgnoreMapper.java (hunk 5 above)
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/ignore/ErroneousCombinationOfSourceAndIgnoreMapper.java]: I'll remove that.. missed that initially..

**Comment by sjaakd** [comment]
test this please

**Comment by filiphr** [comment]
Are the integration tests passing on your side?

Btw, with Travis you don't need to tell it to run the tests. They are always automatically run 😄.

File: processor/src/test/java/org/mapstruct/ap/test/inheritfromconfig/InheritFromConfigTest.java (hunk 2 above)
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/inheritfromconfig/InheritFromConfigTest.java]: Can you check if its complete now.. ?

File: processor/src/main/java/org/mapstruct/ap/internal/prism/MappingInheritanceStrategyPrism.java [hunk 6]
```diff
@@ -25,6 +25,32 @@
  * @author Andreas Gudian
  */
 public enum MappingInheritanceStrategyPrism {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/prism/MappingInheritanceStrategyPrism.java]: Do we need a unit test to make sure that the parameters here are correct? I know they would break the other tests. I can add it later on as well

File: processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java [hunk 7]
```diff
@@ -446,41 +446,48 @@ private void mergeInheritedOptions(SourceMethod method, MapperConfiguration mapp
...
-                    true,
-                    method,
-                    messager,
-                    typeFactory );
-            }
-            else if ( applicableReversePrototypeMethods.size() > 1 ) {
-                messager.printMessage(
-                    method.getExecutable(),
-                    Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
-                    Strings.join( applicablePrototypeMethods, ", " ) );
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java]: I just realized that this was wrong all along 😄. It seems we don't have a test for this error case

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 4 above)
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: This one is still here. We don't need it right?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1255/Issue1255Test.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ */
+@IssueKey("1255")
+@RunWith(AnnotationProcessorTestRunner.class)
+@WithClasses({
+    AbstractA.class,
+    SomeA.class,
+    SomeB.class,
+    SomeMapper.class,
+    SomeMapperConfig.class})
+@WithSingleCompiler( org.mapstruct.ap.testutil.runner.Compiler.JDK )
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1255/Issue1255Test.java]: This should not be used, I supposed you were only using it locally?

Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1255/Issue1255Test.java]: ach.. forgot to remove it..

File: processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java (hunk 7 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java]: true... I'll have a look.. Dunno if I've time left.. Bit tired..

**Comment by filiphr** [comment]
@sjaakd I think that it is fine. Just remove the `@WithSingleCompiler` and the obsolete message and it is good for merging. I'll make an issue for the missing test case, there was none from the get go as well

**Comment by sjaakd** [comment]
> @sjaakd I think that it is fine. Just remove the @WithSingleCompiler and the obsolete message and it is good for merging. I'll make an issue for the missing test case, there was none from the get go as well

I've just added the test case. So I'll squash all the stuff in one commit if CI succeeds.
